proof/invariant-abstract/DetSchedInvs_AI.thy

```diff
@@ -581,26 +581,30 @@ end
 
 \<comment> \<open>Project threads from the kernel heap\<close>
 
-definition tcb_of :: "kernel_object \<rightharpoonup> tcb" where
-  "tcb_of ko \<equiv> case ko of TCB tcb \<Rightarrow> Some tcb | _ \<Rightarrow> None"
-
-lemma tcb_of_simps:
+primrec tcb_of :: "Structures_A.kernel_object \<rightharpoonup> tcb" where
   "tcb_of (TCB tcb) = Some tcb"
-  "tcb_of (SchedContext sc n) = None"
-  "tcb_of (CNode n cnode) = None"
-  "tcb_of (Endpoint ep) = None"
-  "tcb_of (Notification ntfn) = None"
-  "tcb_of (Reply reply) = None"
-  "tcb_of (ArchObj aobj) = None"
-  by (auto simp: tcb_of_def)
+| "tcb_of (Endpoint _) = None"
+| "tcb_of (Notification _) = None"
+| "tcb_of (CNode _ _) = None"
+| "tcb_of (Reply _) = None"
+| "tcb_of (SchedContext _ _) = None"
+| "tcb_of (ArchObj _) = None"
+
+lemma tcb_of_SomeD[dest!]:
```

Comment:
Not sure any more, I might have written the `tcb_of_Some` version below, but I think this could be an equality instead of a `==>`. With an equality, this could then be `[simp]`, and we can probably remove `tcb_of_Some`.

Reply:
Yes, there are equality versions below and I didn't need to copy these ones across. I can add the equalities to the simp set.